lib/ecto/type.ex

```diff
@@ -145,6 +147,13 @@ defmodule Ecto.Type do
   """
   @callback dump(term) :: {:ok, term} | :error
 
+  @doc """
+  Checks if two terms are semantically equal.
+  """
+  @callback equal?(term, term) :: boolean
```

Comment:
other callbacks have `{:ok, term} | :error` return, not sure if we need it here too